Lodash: Improve support for NumericDictionary collections

UselessPickles opened the pull request:
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/DefinitelyTyped/DefinitelyTyped/issues/23723
- [x] Increase the version number in the header if appropriate. (not appropriate)
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. (already exists)

Fixes https://github.com/DefinitelyTyped/DefinitelyTyped/issues/23723, along with other related improvements.

Generally increased support for properly typed results when using lodash with `NumericDictionary` collections (object with numeric key type).

* In methods that use iterators (`forEach`, `map`, etc.), the `key` param of the iterator is now `string` to match the run-time type of keys of objects.
* `omit` now returns exactly the same kind of `Dictionary/NumericDictionary` that was provided as a param.
* Many methods updated to remove overloads to handle `NumericDictionary` explicitly, and now instead rely on more generic handling of `T extends object` where possible.
* Many other methods that previously accepted `Dictionary` param are now updated to also accept `NumericDictionary`.
* Added unit tests for `mapValues`, which uncovered many flaws with signature overloads. Added a few missing signatures, re-ordered as necessary, etc.
* Removed explicit type parameters from many calls to methods in unit tests.
* Several other small improvements/additions to unit tests.

UselessPickles commented:
NOTE: The following unit test failure was happening before I made any changes, so it is unrelated to my changes:

```
Error: D:/DevProjects/github/DefinitelyTyped/types/lodash/lodash-tests.ts:8259:38
ERROR: 8259:38  expect  TypeScript@next compile error:
Argument of type '{ foo: (v: string) => boolean; }' is not assignable to parameter of type 'ConformsPredicateObject<{ foo: any; }>'.
  Types of property 'foo' are incompatible.
    Type '(v: string) => boolean' is not assignable to type '((val: {}) => boolean) | undefined'.
      Type '(v: string) => boolean' is not assignable to type '(val: {}) => boolean'.
        Types of parameters 'v' and 'val' are incompatible.
          Type '{}' is not assignable to type 'string'.
```

UselessPickles commented:
@aj-r Please take a look. You said you had some ideas that would be easier to discuss as comments on a PR.

aj-r commented on types/lodash/common/object.d.ts:
```diff
@@ -2298,50 +2274,93 @@ declare module "../index" {
+         *       but TypeScript cannot select overload signatures based on number vs string index key type.
          */
-        mapValues<T extends object, TKey extends keyof T>(obj: T | null | undefined, iteratee: string): { [P in keyof T]: any };
+        mapValues<T>(obj: Dictionary<T> | NumericDictionary<T> | null | undefined, iteratee: string): Dictionary<T[keyof T]>;
```
I think this should return `Dictionary<any>`. If we reach this overload, it's because `iteratee` does not match `keyof T`. So either they're specifying a property that doesn't exist, in which case the value will be `undefined`, or they're accessing some property that TypeScript doesn't know about, so we have no idea what the type will be.

Don't forget to update the wrapper versions, too.

UselessPickles replied:
I was going to disagree with you for reasons that are not worth fully explaining, because I gathered from reading documentation that a `string` iteratee is not necessarily just a property name, but could also be a property "path" deeper into nested properties of the object. This support of "property paths" means we can't make any assumptions at all about the result type.

aj-r commented on types/lodash/common/object.d.ts:
```diff
@@ -3173,10 +3242,11 @@ declare module "../index" {
     }
 
     interface LoDashImplicitWrapper<TValue> {
+        // NOTE: this overload also works for NumericDictionary and arbitrary object types.
```
Is this note really necessary? If it is, I would put it on the `LoDashStatic` version, not here.

UselessPickles replied:
Oops. That comment is leftover from a different attempt at the signature.  Deleting the comment.

aj-r commented on types/lodash/lodash-tests.ts:
```diff
@@ -4062,29 +4062,52 @@ namespace TestAt {
+    {
+        let result: Array<AbcObject[keyof AbcObject]>;
+
+        result = _.at(abcObject, 'a', ['b'], ['a', 'b']);
```
Testing using `result = x` is an old way of doing it, and it's not very good because it only proves that `x` is *assignable to* `Array<AbcObject[keyof AbcObject]>`. It doesn't prove that the return type is *exactly* `Array<AbcObject[keyof AbcObject]>`. For example, `_.at()` could return `any`, and this test would still pass.

A better way is to use `$ExpectType`:

```ts
_.at(abcObject, 'a', ['b'], ['a', 'b']); // $ExpectType AbcObject[keyof AbcObject][]
```

The downside is that `dtslint` does a dumb string comparison when it sees `$ExpectType`, so your type needs to *exactly match* what `dtslint` is expecting, including whitespace. Also, you'll need to use `[]` instead of `Array<>`.

Also, since `$ExpectType` allows you to remove `let result: x;`, the braces (`{}`) are no longer necessary. i.e. instead of this:

```ts
{
    _.at(abcObject, 'a', ['b'], ['a', 'b']); // $ExpectType AbcObject[keyof AbcObject][]
}
```

just do this:

```ts
_.at(abcObject, 'a', ['b'], ['a', 'b']); // $ExpectType AbcObject[keyof AbcObject][]
```

aj-r replied:
This comment also applies to tests below.

UselessPickles replied:
Yeah; there were a couple bugs I found already that were masked by false positive test cases using  "assignability". I started updating some unit tests and writing new unit tests using $ExpectType comments later on. It's hard to decide when to follow the existing pattern vs when to rework to the better approach. Gotta be careful to avoid going down the rabbit hole that results in every unit test being rewritten and creating an impossibly gigantic PR to review.

I'll work on updating some more of these test cases. Let me know if you have any good advice on how far down the rabbit hole I should go.

aj-r replied:
Agreed. In general, my approach is to clean up tests as I'm touching them. And if I create new ones, obviously create them using the better style. Probably don't worry about the tests where you only removed `<AbcObject>`, though.

Right now, actually, don't go to crazy updating the tests. I just found out that in order to support `ReadonlyArray<T>`, I'm going to need to do some major optimizations to the tests, which means I'll probably end up converting them all to the new style anyway.

aj-r commented on types/lodash/lodash-tests.ts:
```diff
@@ -5335,43 +5347,63 @@ namespace TestForEach {
             collection; // $ExpectType ArrayLike<AbcObject>
         });
+
+        result; // $ExpectType ArrayLike<AbcObject>
```
To simplify this, you should be able to put the `$ExpectType` above `_.forEach`:

```ts
// $ExpectType ArrayLike<AbcObject>
_.forEach(list, (value, index, collection) => {
    value; // $ExpectType AbcObject
    index; // $ExpectType number
    collection; // $ExpectType ArrayLike<AbcObject>
});
```

UselessPickles replied:
I figured that out later in the mapValues tests :). Forgot to go back and simplify these tests.

aj-r commented on types/lodash/lodash-tests.ts:
```diff
@@ -6068,20 +6156,45 @@ namespace TestMap {
     let array: number[] | null | undefined = [] as any;
     let list: _.List<number> | null | undefined = [] as any;
     let obj: any = {};
```
Delete this line

aj-r commented on types/lodash/lodash-tests.ts:
```diff
@@ -6068,20 +6156,45 @@ namespace TestMap {
     let list: _.List<number> | null | undefined = [] as any;
     let obj: any = {};
-    let dictionary: _.Dictionary<number> | null | undefined = obj;
+    let dictionary: _.Dictionary<number> | null | undefined = anything;
```
Use `const` instead of `let`.

Don't bother updating *all* usages of `let`, otherwise this PR will be nearly impossible to review. But if you're changing a line anyways, might as well change it to `const`.

aj-r commented on types/lodash/lodash-tests.ts:
```diff
@@ -6068,20 +6156,45 @@ namespace TestMap {
 
     let listIterator: (value: number, index: number, collection: _.List<number>) => AbcObject = (value: number, index: number, collection: _.List<number>) => ({ a: 1, b: "", c: true });
     let dictionaryIterator: (value: number, key: string, collection: _.Dictionary<number>) => AbcObject = (value: number, key: string, collection: _.Dictionary<number>) => ({ a: 1, b: "", c: true });
+    let numericDictionaryIterator: (value: number, key: string, collection: _.NumericDictionary<number>) => AbcObject = (value: number, key: string, collection: _.NumericDictionary<number>) => ({ a: 1, b: "", c: true });
```
This style was written by me when I was young and foolish. A simpler way would be:

```ts
const numericDictionaryIterator = (value: number, key: string, collection: _.NumericDictionary<number>): AbcObject => ({ a: 1, b: "", c: true });
```

Or better yet, don't declare a `numericDictionaryIterator` at all, and instead write the function inline (It looks like you've already done that below):

```ts
// $ExpectType ???
_.map(numericDictionary, (value, index, collection) => {
    value; // $ExpectType number
    index; // $ExpectType string
    collection; // $ExpectType NumericDictionary<number>
});
```

aj-r commented on types/lodash/lodash-tests.ts:
```diff
@@ -6068,20 +6156,45 @@ namespace TestMap {
     {
         _.map(array);  // $ExpectType number[]
         _.map(array, listIterator);  // $ExpectType AbcObject[]
+        _.map(array, (value, index, collection) => {
```
Add `$ExpectType` above this line

aj-r commented on types/lodash/lodash-tests.ts:
```diff
@@ -6068,20 +6156,45 @@ namespace TestMap {
 
     {
         _.map(array);  // $ExpectType number[]
         _.map(array, listIterator);  // $ExpectType AbcObject[]
```
I feel like this line isn't necessary anymore - it's a worse version of the test you do on the next line.

aj-r commented on types/lodash/lodash-tests.ts:
```diff
@@ -6330,27 +6489,27 @@ namespace TestPartition {
 //
-//         result = _(array).map<AbcObject>('d.0.b');
-//         result = _(array).map<AbcObject>(['d', 0, 'b']);
+//         result = _(array).map('d.0.b');
```
I think you should actually remove these lines, and add proper `_.map` tests above (line ~6200). You can use these tests as inspiration before you delete them :)

aj-r replied:
Yeah I would ignore that for now. If you want, you can review it later and create a separate PR (or at least create an issue so someone else can create a PR).

I'm not 100% convinced that `_.map` should accept an array of strings. If it does, then all `iteratees` should accept an array of strings. Again, this is a problem for a separate PR, so probably delete those tests for now.

UselessPickles replied:
I just now reverted my change to support PropertyPath in `_.map`.

aj-r commented on types/lodash/lodash-tests.ts:
```diff
@@ -10488,53 +10627,51 @@ result = <DefaultsDeepResult>_(TestDefaultsDeepObject).defaultsDeep(TestDefaults
 // _.entries
 namespace TestEntries {
-    let object: _.Dictionary<string> = {};
+    let dictionary: _.Dictionary<number> = {};
```
use `const`

aj-r commented on types/lodash/lodash-tests.ts:
```diff
@@ -11556,65 +11693,332 @@ namespace TestMapKeys {
+    }
+}
+
+// _.mapValues
```
Wow, I didn't even realize we were missing `mapValues` tests.

Looks really good. Although probably make a few adjustments based on my previous comments.

aj-r replied:
Also, I feel like all of the braces (`{}`) aren't necessary since you don't ever do `let result: x;`. For other tests, the braces are needed to keep scopes separate. But that's not an issue here, so I would just use line breaks instead of braces to logically group things.

aj-r commented:
Github claims that Travis passed, but I see these errors:

- in Bookshelf: `Namespace '_' has no exported member 'DictionaryIterator'.` (did you remove the `DictionaryIterator` type? Might need to add it back for backwards compatibility.)
- GC (out of memory?) error when testing with TS 2.6

The GC errors are really annoying. The only way I know to find the real problem is to delete parts of the tests until you no longer get the GC error. Then it will tell you what the real error is (usually some error that only happens in TS 2.3/2.4).

I'm in the process of diagnosing a GC error related to `ReadonlyArray`, and in the process I found some of the errors for you :)
- `unionBy`: Errors are occurring for the last two 'a' overloads in TS 2.3. It's complaining that it can't determine the type of `T`. Looks like you'll have to manually pass the type parameter.
- `forEach`: The `value; // $ExpectType X` lines are failing for your NumericDictionaries in TS 2.4. Apparently TS 2.4 gets a type of `any` in that case. I can live with that, so let's just remove the `value; // $ExpectType X` lines.
- `some`: `Argument of type '{ a: number; }' is not assignable to parameter of type 'string | ... | object | ...'. Object literal may only specify known properties, and 'a' does not exist...`. Not sure how this broke. Looks like you need to add the `NumericDictionary` overloads back for `_.some`, but I'm baffled as to why. Also remove the `<{ a: number}>` in the tests.

UselessPickles commented on types/lodash/common/common.d.ts:
```diff
@@ -230,6 +222,17 @@ declare module "../index" {
     }
 
+    // Crazy typedef needed get _.omit to work properly with Dictionary and NumericDictionary
+    type AnyKindOfDictionary =
```
That does work. I clearly don't quite understand the type system of TypeScript yet. Every time I think I got it, I run into something like this.

UselessPickles commented on types/lodash/common/collection.d.ts:
```diff
@@ -1788,46 +1625,48 @@ declare module "../index" {
          * @see _.map
          */
-        map<T>(collection: List<T> | Dictionary<T> | null | undefined): T[];
+        map<T>(collection: List<T> | Dictionary<T> | NumericDictionary<T> | null | undefined): T[];
```
Without the `| NumericDictionary<T>`, it incorrectly infers type `T` to be `any` when passing a `NumericDictionary` as the param.

aj-r commented:
I just noticed the [???] boxes in your original description. You don't need to do anything for those.

> [???] Increase the version number in the header if appropriate.

It's not appropriate in this situation. That's only necessary if we're updating the types for a new version of lodash, which we're not.

> [???] If you are making substantial changes, consider adding a tslint.json containing { "extends": "dtslint/dt.json" }.

This has already been done :)

UselessPickles commented:
I think I took care of everything except the GC error stuff. Can you give me some guidance on how to verify my changes for that, rather than blindly making the changes you suggest and hoping I interpreted you correctly?

aj-r commented on types/lodash/common/collection.d.ts:
```diff
@@ -2943,14 +2599,6 @@ declare module "../index" {
-        /**
-         * @see _.some
-         */
-        some<T>(
```
For reasons unknown to me, you need to add this overload back to make the tests work in TS 2.4. (same with wrappers)

UselessPickles replied:
What I found that seems to fix this is not adding back this overload, but changing the definitions of:

```ts
    type ObjectIteratee<TObject> = ObjectIterator<TObject, NotVoid> | string | [string, any] | PartialDeep<TObject[keyof TObject]>;
    type ObjectIterateeCustom<TObject, TResult> = ObjectIterator<TObject, TResult> | string | object | [string, any] | PartialDeep<TObject[keyof TObject]>;
```

to 
```ts
    type ObjectIteratee<TObject> = ObjectIterator<TObject, NotVoid> | string | [string, any] | PartialDeep<TObject>;
    type ObjectIterateeCustom<TObject, TResult> = ObjectIterator<TObject, TResult> | string | object | [string, any] | PartialDeep<TObject>;
```

(focus on the `PartialDeep` part)

I don't understand the purpose of this well enough to know whether this change is correct, or the unit test's usage of `{ a: 42 }` as an iterator is an incorrect unit test, or neither, or both.

It was the `sampleObject` tests of _.some that were failing for me and were fixed by the above change:

```ts
        result = _.some(sampleObject, {a: 42});
```

UselessPickles replied:
After looking at this some more, I think the test that is failing is actually invalid. It's iterating over properties of `SampleObject`, which are of type `number`, `string` and `boolean`, so the `{a: 42}` iterator doesn't make any sense at all - none of the properties of `SampleObject` are of a type that have a property `a`. 

Do you agree? Should I simply delete the tests that use the `{a: 42}` iterator against a collection of type `SampleObject`?

aj-r replied:
Yeah the test looks invalid. I think PatialDeep needs to stay as it was before (`TObject [keyof TObjet]`) but you should remove the `| object` type (that's my mistake from a while ago). Feel free to remove the test.

aj-r commented on types/lodash/lodash-tests.ts:
```diff
@@ -11556,65 +11606,274 @@ namespace TestMapKeys {
+        _.mapKeys(numericDictionary); // $ExpectType Dictionary<AbcObject>
+        _.mapKeys(numericDictionary, numericDictionaryIterator); // $ExpectType Dictionary<AbcObject>
+        _.mapKeys(numericDictionary, ''); // $ExpectType Dictionary<AbcObject>
+        _.mapKeys(numericDictionary, {}); // $ExpectType Dictionary<AbcObject>
```
These tests don't seem to work in TS 2.4 - it gets a return type of `Dictionary<any>`. I guess it's best to just remove these tests.

UselessPickles commented:
I really don't know how to proceed with this whole "GC error" thing, because I don't even see it when I run `npm run lint lodash` on my machine. This is the only error I get (which also existed before I started making any changes at all):

```

Error: D:/DevProjects/github/DefinitelyTyped/types/lodash/lodash-tests.ts:8172:38
ERROR: 8172:38  expect  TypeScript@next compile error:
Argument of type '{ foo: (v: string) => boolean; }' is not assignable to parameter of type 'ConformsPredicateObject<{ foo: any; }>'.
  Types of property 'foo' are incompatible.
    Type '(v: string) => boolean' is not assignable to type '((val: {}) => boolean) | undefined'.
      Type '(v: string) => boolean' is not assignable to type '(val: {}) => boolean'.
        Types of parameters 'v' and 'val' are incompatible.
          Type '{}' is not assignable to type 'string'.
```
I'm uncomfortable with making changes that I cannot validate myself.

Is there something else I have to do to get the unit tests to run with different versions of TypeScript?

aj-r commented:
Weird, I don't see that error on my machine. Try deleting your node_modules
folder and running mom install again?

On Mar 5, 2018 9:13 PM, "Jeff Lau" <notifications@github.com> wrote:

> I really don't know how to proceed with this whole "GC error" thing,
> because I don't even see it when I run npm run lint lodash on my machine.
> This is the only error I get (which also existed before I started making
> any changes at all):
>
>
> Error: D:/DevProjects/github/DefinitelyTyped/types/lodash/lodash-tests.ts:8172:38
> ERROR: 8172:38  expect  TypeScript@next compile error:
> Argument of type '{ foo: (v: string) => boolean; }' is not assignable to parameter of type 'ConformsPredicateObject<{ foo: any; }>'.
>   Types of property 'foo' are incompatible.
>     Type '(v: string) => boolean' is not assignable to type '((val: {}) => boolean) | undefined'.
>       Type '(v: string) => boolean' is not assignable to type '(val: {}) => boolean'.
>         Types of parameters 'v' and 'val' are incompatible.
>           Type '{}' is not assignable to type 'string'.
>
> I'm uncomfortable with making changes that I cannot validate myself.
>
> Is there something else I have to do to get the unit tests to run with
> different versions of TypeScript?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/DefinitelyTyped/DefinitelyTyped/pull/24052#issuecomment-370636967>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGpes_zWIXmbQ_eNFiJGohvAgPo5bBX-ks5tbfDXgaJpZM4SavvM>
> .
>

UselessPickles commented:
That worked. Now I get the GC error. I'll try to make some progress on this tomorrow night.

UselessPickles commented:
Some thoughts on the "GC errors" and failures with older versions of TypeScript:
* Failures in older versions of TypeScript do not seem to be causing the GC errors from what I'm seeing. The tests are run in reverse order (newest versions first). Initially, tests are failing due to a memory allocation failure (FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory) while testing with TypeScript 2.6.  Deleting a large chunk of tests allows 2.6 to complete successfully, but then 2.5 fails with a similar allocation error. I'm thinking that it is the out-of-memory errors in newer versions of TS masking failures in older versions, rather than failures in older versions causing the out-of-memory errors. Is this simply because ALL of the tests are in one giant file? Would splitting the file into several smaller files fix this?
* For tests that fail in older versions of TypeScript due to incorrectly inferred types, etc., it would be nice to split up the tests in a way that the full functionality can still be tested with newer versions of TypeScript. I just experimented with this concept in one of my projects. Check it out: https://github.com/UselessPickles/ts-enum-util/tree/master/type_tests

aj-r commented:
Yeah dtslint seems to run out of memory the more tests it runs. It's possible that dtslint could improve this but I'm not sure. You are right - the GC error is just masking errors in older typescript versions.

I would love to split the tests into multiple files, but unfortunately DefinitelyTyped does not allow it. If you try to add more test files, Travis fails with an error saying that the test file must be named `lodash-tests.ts` and must be placed at the root. I've also tried having `lodash-tests` import other test files, but then Travis fails saying that relative imports are not allowed.

@andy-ms Is there anything we can do to `dtslint` to make it more memory-efficient? Or at least allow multiple test files?

andy-ms commented:
You can have multiple test files in the `test` directory. I don't think that will improve efficiency though.
Assuming that `dtslint`'s memory usage is due to asking typescript to type-check everything, I don't think there's much that can be done -- `typescript@2.2` isn't going to get more memory efficient.
Note that dtslint doesn't normally test in all typescript versions -- it only does this if there was a failure in the earliest version, then works backwards to find the latest version that fails.

thorn0 commented:
Sounds like it could help if the test files were type checked one by one in separate invocations of TS.

andy-ms commented:
Do you mean launch a separate process for each test file and ts version combination we test? That would get pretty slow -- we used to launch a separate process per *package* and that was enough that travis would time out.

UselessPickles commented:
@andy-ms  Thanks for the clarification of how/when dtslint tests all versions of TS working backward. That makes more sense now as to how there may be some situations where fixing a failure in an earlier TS version can actually "fix" the out-of-memory error (by preventing dtslint from running against all later versions)

aj-r commented:
@UselessPickles any success fixing the errors? It would be good to merge this soon. I don't want to start working on the optimizations until this is done, otherwise there will be major merge conflicts.

UselessPickles commented on types/lodash/lodash-tests.ts:
```diff
@@ -2628,157 +2628,157 @@ namespace TestUnionBy {
+        result = _.unionBy(array, list, array, 'a');
+        result = _.unionBy(array, array, list, array, 'a');
+        result = _.unionBy(array, list, array, list, array, 'a');
+        result = _.unionBy(array, array, list, array, list, array, 'a');
```
Done

UselessPickles commented:
Another failure I don't know how to fix, for TS 2.4:

```ts
    _.pick(obj2, 'a', 'b'); // $ExpectType Pick<AbcObject, "a" | "b">
```

```
ERROR: 846:5  expect  Expected type to be:
  Pick<AbcObject, "a" | "b">
got:
  Pick<AbcObject, "a" | "b" | "c">
```

UselessPickles commented:
Looks like `_.some` and `_.pick` (as mentioned in other comments) are the last two remaining failures.

UselessPickles commented:
I'll try to finish this up tonight.

UselessPickles commented:
@aj-r I think it might finally be done. I commented out the broken _.pick tests, deleted in the invalid _.some tests, and deleted `object` from the definition of `ObjectIterateeCustom`. 

Now `npm run lint lodash` passes. I hope I didn't break anything.

UselessPickles commented:
@aj-r Travis is not happy :(. I looked at the errors in the logs, but it looks completely unrelated to lodash to me. I need help interpreting the failure.

aj-r commented:
Looks like the error is not caused by this PR. @andy-ms I think this can be merged anyways.

andy-ms commented:
Thanks, that issue is Microsoft/TypeScript#22603
